app/models/attachment_reference.rb

```diff
@@ -0,0 +1,57 @@
+class AttachmentReference < ActiveRecord::Base
+  after_initialize :set_expiry_at, if: :new_record?
```

Comment:
Actually all the logic was inside before_save,  if there is a owner, `expiry_at` will be reset there.  But override `owner=` method might be better, in case callbacks are skipped.